src/simple_server/core.clj

```diff
@@ -2,38 +2,72 @@
             [compojure.coercions :refer [as-int]]
             [ring.middleware.defaults :as middleware]
+            [ring.middleware.cookies :refer [wrap-cookies]]
 
             [ring.mock.request :as mock]
-            [simple-server.simple-game :as game]))
+            [simple-server.simple-game :as game]
+            [clojure.edn :as edn]))
 
 ;;; Finally, let us truly separate concerns between our "application code"
 ;;; and our "http code".  Our game now lives in its own namespace, and
 ;;; is fully testable independent of our "presentation layer".
 
-(defn new-game-handler []
-  (when (game/new-game!)
-    (response "OK- start guessing at /guess")))
+(def users-info
+  "Store all users' login information"
+  (atom {}))
 
-(defn guess-handler [guess]
-  (condp = (game/guess-answer guess)
-    nil       (-> (response  "You need to supply a guess with /guess?guess=N")
-                  (status 400))
-    :game-over (response  "Congratulations! You win!")
-    :too-low   (response "Too low.")
-    :too-high  (response  "Too high.")))
+(defn login?
+  "Check if a user has logged in"
+  [user-collection user]
+  (@user-collection user))
+
+(defn login
+  [user-collection user]
+  (swap! user-collection assoc user true))
+
+(defn login-handler
+  [username]
+  (login users-info username)
+  (-> (response (format "Welcome to the guessing game! %s" username))
+      (assoc :cookies {"session_id" {:value username}})))
+
+(defn new-game-handler [username]
+  (when (game/new-game! username)
+    (response (format "OK- start guessing at /guess. Current user: %s" username))))
+
+(defn guess-handler [username guess]
+  (condp = (game/guess-answer username guess)
+    nil               (-> (response  "You need to supply a guess with /guess?guess=N")
+                          (status 400))
+    :game-not-started (-> (response "You need to start the game with /new-game first.")
+                          (status 400))
+    :game-over        (response (format "Sorry, %s, you have guessed 5 times. Game over." username))
+    :win              (response (format "Congratulations %s! You win!" username))
+    :too-low          (response (format "Too low and you have %s chances left." (- 5 (game/get-times game/game-in-progress username))))
+    :too-high         (response (format "Too high and you have %s chances left." (- 5 (game/get-times game/game-in-progress username))))))
 
 (defroutes game-routes
-  (GET "/new-game" []                 (new-game-handler))
-  (GET "/guess"    [guess :<< as-int] (guess-handler guess))
-  (ANY "*"         []                 (not-found "Sorry, No such URI on this server!")))
+  (POST "/login"    [username]                                                           (login-handler username))
+  (POST "/new-game" {{{username :value} "session_id"} :cookies}                          (if (login? users-info username) (new-game-handler username) (response "Please login first.")))
```

Comment:
This is the bit I don't like much.

It seems poor taste to do all this computation inline
on the POST,PUT lines, and even worse to have basically
the same computation repeated twice.

Can you think of another way to accomplish this?  Middleware?